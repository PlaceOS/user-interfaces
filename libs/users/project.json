{
    "name": "users",
    "$schema": "../../node_modules/nx/schemas/project-schema.json",
    "projectType": "library",
    "sourceRoot": "libs/users/src",
    "prefix": "placeos",
    "targets": {
        "lint": {
            "executor": "@nx/eslint:lint",
            "outputs": ["{options.outputFile}"]
        },
        "test": {
            "executor": "@nx/jest:jest",
            "outputs": ["{workspaceRoot}/coverage/libs/users"],
            "options": {
                "jestConfig": "libs/users/jest.config.ts"
            }
        }
    },
    "tags": []
}
