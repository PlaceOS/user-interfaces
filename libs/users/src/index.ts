export * from './lib/users.module';
export * from './lib/new-user-modal.component';
export * from './lib/find-availability-modal/find-availability-modal.component';
export * from './lib/user-form.component';
export * from './lib/location.class';
export * from './lib/user.class';
export * from './lib/user.utilities';
export * from './lib/users.module';
export * from './lib/guests.fn';
export * from './lib/staff.fn';
