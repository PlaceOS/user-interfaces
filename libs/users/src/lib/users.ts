export * from './new-user-modal.component';
export * from './user-form.component';
export * from './location.class';
export * from './user.class';
export * from './user.utilities';
export * from './users.module';
export * from './guests.fn';
export * from './staff.fn';
