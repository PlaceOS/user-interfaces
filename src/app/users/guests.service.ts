import { Injectable } from '@angular/core';
import { updateMetadata } from '@placeos/ts-client';

import { GuestUser } from './user.class';
import { CalendarEvent } from '../events/event.class';
import { BaseAPIService } from '../common/base.service';
import { HashMap } from '../common/types';

@Injectable({
    providedIn: 'root',
})
export class GuestsService extends BaseAPIService<GuestUser> {
    constructor() {
        super();
        this._name = 'Guest';
        this._api_route = 'guests';
    }

    protected process(raw_data: HashMap): GuestUser {
        return new GuestUser(raw_data);
    }

    /**
     * Update metadata associated with guest
     * @param email Email of the guest
     * @param data New state of the guest's metadata
     */
    public updateMetadata(email: string, data: HashMap | any[]) {
        return updateMetadata(email, { id: email, name: 'guest-metadata', details: data });
    }

    /**
     * Get upcoming meetings for a guest
     * @param id Email Address of the guest
     */
    public meetings(id: string): Promise<CalendarEvent[]> {
        return this.task(id, 'meetings', {}, 'get');
    }
}
