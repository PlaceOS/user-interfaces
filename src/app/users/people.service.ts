import { Injectable } from '@angular/core';
import { BaseAPIService } from '../common/base.service';
import { User } from '../users/user.class';

@Injectable({
    providedIn: 'root',
})
export class PeopleService extends BaseAPIService<User> {
    constructor() {
        super();
        this._name = 'People';
        this._api_route = 'people';
        this._compare = (a, b) => !a.id.localeCompare(b.id) || !a.email.localeCompare(b.email);
    }

    public async search(search: string) {
        return this.query({ q: search });
    }

    public locate(email: string) {
        return this.show(email);
    }
}
