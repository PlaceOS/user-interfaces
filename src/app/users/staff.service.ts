import { Injectable } from '@angular/core';
import { onlineState, show } from '@placeos/ts-client';
import { first } from 'rxjs/operators';

import { StaffUser } from './user.class';
import { BaseAPIService } from '../common/base.service';
import { HashMap } from '../common/types';

@Injectable({
    providedIn: 'root',
})
export class StaffService extends BaseAPIService<StaffUser> {
    /** Currently logged in user */
    private _active_user: StaffUser;

    /** Currently logged in user */
    public get current(): StaffUser {
        return this._active_user;
    }

    constructor() {
        super();
        this._name = 'Staff';
        this._api_route = 'people';
        this._active_user = new StaffUser({ name: 'Local User' });
        onlineState()
            .pipe(first((_) => _))
            .subscribe(() => {
                this.timeout('init', () => this.init());
            });
    }

    /** Load current user */
    public async load() {
        const user = await show('current', {}, this.format, 'users').toPromise();
        if (user) {
            this._active_user = new StaffUser({ ...user, is_logged_in: true });
        }
    }

    public format(item) {
        return item;
    }

    public process(raw_data: HashMap) {
        return new StaffUser(raw_data);
    }

    /**
     * Get location of a staff member
     * @param id Email address of the staff member
     */
    public locate(id: string): Promise<StaffUser> {
        return this.show(id);
    }
}
