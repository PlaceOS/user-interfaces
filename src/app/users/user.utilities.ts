
import { FormControl, FormGroup, Validators } from '@angular/forms';

import { HashMap } from '../common/types';
import { unique } from '../common/general';
import { User } from './user.class';

import * as faker from 'faker';

let USER_COUNT = 0;

export let USER_DOMAIN = 'acaprojects.com';

const USER_EMAILS: string[] = [];

/**
 * Generate raw mock data for a user
 * @param id Forced ID for the mock
 * @param name Forced name for the user
 * @param external Whether user is external of the organisation
 */
export function generateMockUser(id?: string, name?: string, external?: boolean): HashMap {
    if (!id) {
        id = `user-${USER_COUNT++}`;
    }
    if (!name) {
        name = `${faker.name.firstName()} ${faker.name.lastName()}`;
    }
    if (external === undefined) {
        external = !((Math.random() * 99999) % 2);
    }
    const organisation = external ? faker.company.companyName() : USER_DOMAIN.split('.')[0];
    let delegates: string[] = [];
    const delegate_count = Math.min(Math.random() * 4 + 1, USER_EMAILS.length);
    for (let i = 0; i < delegate_count; i++) {
        delegates.push(USER_EMAILS[Math.floor(Math.random() * USER_EMAILS.length)]);
    }
    delegates = unique(delegates);
    const email = `${name.split(' ').join('.').toLowerCase()}@${
        external ? 'not-' : ''
    }${USER_DOMAIN}`;
    USER_EMAILS.push(email);
    return {
        id,
        name,
        first_name: name.split(' ')[0],
        last_name: name.split(' ')[1],
        email,
        phone: faker.phone.phoneNumber(),
        visitor: external,
        organisation: {
            id: organisation.split(' ').join('.').toLowerCase(),
            name: organisation,
        },
        department: faker.commerce.department(),
        staff_code: Math.floor(Math.random() * 99999),
        delegates,
        image: faker.image.avatar(),
    };
}

/**
 * Generate form fields for the given user
 * @param user User to generate form for
 */
export function generateUserForm(user: User): FormGroup {
    if (!user) {
        throw Error('No user passed');
    }

    const fields: HashMap<FormControl> = {
        name: new FormControl(user.name || ''),
        email: new FormControl(user.email || '', [Validators.email, Validators.required]),
        organisation: new FormControl(user.organisation || '', Validators.required),
        phone: new FormControl(user.phone || ''),
        assistance_required: new FormControl(user.assistance_required || false),
    };

    // Generate form group for booking item
    const form = new FormGroup(fields);
    return form;
}
