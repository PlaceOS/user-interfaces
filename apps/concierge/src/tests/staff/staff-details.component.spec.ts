import { MatTooltipModule } from '@angular/material/tooltip';
import { createComponentFactory, Spectator } from '@ngneat/spectator/jest';
import {
    ActionIconComponent,
    IconComponent,
    UserAvatarComponent,
} from '@placeos/components';
import { StaffUser } from '@placeos/users';
import { MockComponent } from 'ng-mocks';
import { StaffDetailsComponent } from '../../app/staff/staff-details.component';
import { StaffStateService } from '../../app/staff/staff-state.service';

describe('StaffDetailsComponent', () => {
    let spectator: Spectator<StaffDetailsComponent>;
    const createComponent = createComponentFactory({
        component: StaffDetailsComponent,
        providers: [
            {
                provide: StaffStateService,
                useValue: {
                    checkin: jest.fn(),
                    checkout: jest.fn(),
                },
            },
        ],
        declarations: [
            MockComponent(ActionIconComponent),
            MockComponent(IconComponent),
            MockComponent(UserAvatarComponent),
        ],
        imports: [MatTooltipModule],
    });

    beforeEach(() => (spectator = createComponent()));

    it('should create component', () => {
        expect(spectator.component).toBeTruthy();
    });

    it('should match snapshot', () => {
        expect(spectator.element).toMatchSnapshot();
        spectator.setInput({
            user: new StaffUser({ name: 'Jim', email: 'jim@test.com' }),
        });
        spectator.detectChanges();
        expect(spectator.element).toMatchSnapshot();
    });

    it.todo('should allow user to checkin');
    it.todo('should allow user to checkout');
});
