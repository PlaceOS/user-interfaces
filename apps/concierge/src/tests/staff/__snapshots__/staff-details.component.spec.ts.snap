// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StaffDetailsComponent should match snapshot 1`] = `
<div
  id="root1"
>
  
</div>
`;

exports[`StaffDetailsComponent should match snapshot 2`] = `
<div
  id="root1"
>
  <div
    class="w-full flex items-center px-4 py-2 bg-white border-b border-gray-200"
  >
    <a-user-avatar />
    <div
      class="flex flex-col flex-1"
    >
      <div
        class="px-2"
      >
        Jim
      </div>
      <div
        class="px-2 text-xs text-opacity-50"
      >
         jim@test.com 
      </div>
    </div>
    
    <div
      class="flex items-center"
    >
      <action-icon
        class="mat-tooltip-trigger"
      />
      
      <a
        class="mat-tooltip-trigger"
        href="mailto:jim@test.com"
        mat-icon-button=""
        mattooltip="Email Staff"
      >
        <app-icon>
          email
        </app-icon>
      </a>
      
      <a
        class="mat-tooltip-trigger"
        href="tel:"
        mat-icon-button=""
        mattooltip="Phone Staff"
      >
        <app-icon>
          call
        </app-icon>
      </a>
      
    </div>
  </div>
  
</div>
`;
